src/test/java/io/numaproj/pulsar/consumer/PulsarSourceTest.java

```
@@ -0,0 +1,227 @@
+    }
+
+    /**
+     * Test the ack method when there is a message to be acknowledged.
+     */
+    @Test
+    public void testAckSuccessful() throws PulsarClientException {
+        // Create a dummy message to acknowledge.
+        org.apache.pulsar.client.api.Message<byte[]> msg = mock(org.apache.pulsar.client.api.Message.class);
+        MessageId msgId = mock(MessageId.class);
+        when(msgId.toString()).thenReturn("ackMsg");
+        when(msg.getMessageId()).thenReturn(msgId);
+        when(msg.getValue()).thenReturn("AckPayload".getBytes(StandardCharsets.UTF_8));
+
+        // Insert the dummy message into the messagesToAck map.
+        java.util.Map<String, org.apache.pulsar.client.api.Message<byte[]>> messagesToAck = (java.util.Map<String, org.apache.pulsar.client.api.Message<byte[]>>) ReflectionTestUtils
+                .getField(pulsarSource, "messagesToAck");
+        messagesToAck.clear();
+        messagesToAck.put("ackMsg", msg);
+
+        // Stub consumerManager to return consumerMock for the ack call.
+        when(consumerManagerMock.getOrCreateConsumer(0, 0)).thenReturn(consumerMock);
+
+        // Create a fake AckRequest with an offset corresponding to the message id.
+        AckRequest ackRequest = new AckRequest() {
+            @Override
+            public java.util.List<Offset> getOffsets() {
+                return Collections.singletonList(new Offset("ackMsg".getBytes(StandardCharsets.UTF_8)));
+            }
+        };
+
+        pulsarSource.ack(ackRequest);
+
+        // Verify that consumer.acknowledge is called on the message.
+        verify(consumerMock, times(1)).acknowledge(msg);
+        // Verify that the messagesToAck map is now empty.
+        assertFalse(messagesToAck.containsKey("ackMsg"));
+    }
+
+    /**
+     * Test the ack method when the offset does not exist in messagesToAck.
+     */
+    @Test
+    public void testAckNoMatchingMessage() throws PulsarClientException {
```

please verify the exception is thrown.

---

Currently my code also just logs an error message indicating the mismatch and then returns early without processing any of the acknowledgments. So, if the key sets don't match, it doesn't throw an exception—it just logs the error and exits the method. But now looking at it again, I feel like I should throw an error. I also took a look at Kafka Java, and I don't believe an error is thrown right away but I might be wrong! Let me know if you want to sync about this.

---

Got it, if so, can we remove the throws `PulsarClientException` statement?

---

and inside the test, let's verify `PulsarClientException` is never thrown. - same for my other comment.

---

i get an error in my ide if i remove it because getOrCreateConsumer is declared with "throws PulsarClientException", and we need it in the method signaure because creating the new consumer can throw that exception. so even though i'm not actually calling the method the compiler still sees the method signature and requires me to handle or declare the checked exception but i agree that it is pretty misleading.  

would it make more sense to just add a try catch in the method and remove it from the method signature? and yes i can cerify the exception is never thrown for all the methods.

---

Yes, `try catch and fail` to verify no unexpected exceptions. We should exercise caution when a unit test has a throws statement in the function signature, because if the exception is thrown before verification, the test could silently pass, false positive.

---

can we apply to source test too?